src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_amanitar.cpp

```diff
@@ -19,218 +19,298 @@
+
+        if (TempSummon* summon = me->SummonCreature(roll_chance_i(40) ? NPC_HEALTHY_MUSHROOM : NPC_POISONOUS_MUSHROOM, MushroomPositions[pos], TEMPSUMMON_CORPSE_TIMED_DESPAWN, 4000))
+            summon->AI()->SetData(DATA_POSITION, pos);
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-                events.Update(diff);
+        events.Update(diff);
 
-                if (me->HasUnitState(UNIT_STATE_CASTING))
-                    return;
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-                while (uint32 eventId = events.ExecuteEvent())
-                {
-                    switch (eventId)
+        while (uint32 eventId = events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_SPAWN:
+                    for (uint32 i = 0; i < MAX_MUSHROOMS; i++)
+                        SpawnMushroom(i);
+                    break;
+                case EVENT_MINI:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 0.0f, true, true, -SPELL_MINI))
                     {
-                        case EVENT_SPAWN:
-                            SpawnAdds();
-                            events.ScheduleEvent(EVENT_SPAWN, 20 * IN_MILLISECONDS);
-                            break;
-                        case EVENT_MINI:
-                            DoCast(SPELL_MINI);
-                            events.ScheduleEvent(EVENT_MINI, urand(25, 30) * IN_MILLISECONDS);
-                            break;
-                        case EVENT_ROOT:
-                            DoCast(SelectTarget(SELECT_TARGET_RANDOM, 0, 100, true), SPELL_ENTANGLING_ROOTS, true);
-                            events.ScheduleEvent(EVENT_ROOT, urand(10, 15) * IN_MILLISECONDS);
-                            break;
-                        case EVENT_BASH:
-                            DoCastVictim(SPELL_BASH);
-                            events.ScheduleEvent(EVENT_BASH, urand(7, 12) * IN_MILLISECONDS);
-                            break;
-                        case EVENT_BOLT:
-                            DoCast(SelectTarget(SELECT_TARGET_RANDOM, 0, 100, true), SPELL_VENOM_BOLT_VOLLEY, true);
-                            events.ScheduleEvent(EVENT_BOLT, urand(18, 22) * IN_MILLISECONDS);
-                            break;
-                        default:
-                            break;
+                        DoCastAOE(SPELL_MINI);
+                        events.Repeat(Seconds(30));
                     }
-
-                    if (me->HasUnitState(UNIT_STATE_CASTING))
-                        return;
-                }
-                DoMeleeAttackIfReady();
+                    else
+                        events.Repeat(Seconds(1));
+                    break;
+                case EVENT_ROOT:
+                    DoCast(SelectTarget(SELECT_TARGET_RANDOM, 1, 100.0f, true), SPELL_ENTANGLING_ROOTS, true);
```

Comment:
better use the structure
```c++
if (target)
    cast(target)
```
cast can be called with null target, and in this cases you dont want a cast with null targets

Reply:
following ur comment on IRC

a spell can be created with null target - it will later on fail on target selection
so, why not saving some runtime?